cpp/cograph_recognition/cograph/Recognition.cpp

```diff
@@ -0,0 +1,233 @@
+#include <recognition/CographRecognition.hpp>
+#include <list>
+#include <set>
+
+#include <graph/GraphTools.hpp>
+namespace Koala {
+enum class Type{
+ZEROONE,
+VERTEX
+};
+enum class Marked{
+    UNMARKED,
+    MARKED,
+    MARKED_AND_UNMARKED
+}
+class CoNode{
+    private:
+    Type type;
+    int number;
+    Marked marked;
+    int md, d;
+    //md is the current number of children, which have been both "marked" and "unmarked"
+    //d is the current number of children
+
+    CoNode* head_of_list_of_children;
+    CoNode *next, *prev;//in list of children of its parent
+    CoNode *parent;
+    vector<CoNode>outEdges;//neighbours of cur vertex in G
+    public:
+    CoNode(Type type, int number): type(type), number(number),marked(Marked::UNMARKED),md(0),d(0),head_of_list_of_children(nullptr),
+    next(nullptr),prev(nullptr), parent(nullptr),outEdges(vector<CoNode>(0)){
+
+    }
+    void addchild(CoNode *x){
+        if(head_of_list_of_children == nullptr){
+            head_of_list_of_children = x;
+            x -> prev = x -> next = nullptr;
+        } else{
+            head_of_list_of_children -> prev = x;
+            x -> next = head_of_list_of_children;
+            x -> prev = nullptr;
+            head_of_list_of_children = x;
+        }
+        x -> parent = this;
+        d++;
+    }
+    void setoutEdges(vector<CoNode>outEdges){
+        this.outEdges = outEdges;
+    }
+    vector<CoNode> getoutEdges(){
+        return outEdges;
+    }
+    CoNode* getParent(){
+        return parent;
+    }
+    CoNode* setParent(CoNode *parent){
+    this.parent = parent;
+    }
+    void set_md(int md){
+        this.md = md;
+    }
+    void inc_md(){
+        this.md++;
+    }
+    int get_md(){
+        return md;
+    }
+    void set_d(int d){
+            this.d = d;
+        }
+        void inc_d(){
+            this.d++;
+        }
+        int get_d(){
+            return d;
+        }
+    Marked Marked_or_not(){
+        return marked;
+    }
+    void unmark_for_new_iteration(){
+    marked = Marked::UNMARKED;
+    md = 0;
+    }
+    void mark(){
+    marked = Marked::MARKED;}
+    void unmark(){
+        marked = Marked::MARKED_AND_UNMARKED;
+    }
+}
+class CoTree{
+    private:
+    CoNode* root;
+    public:
+    CoTree(CoNode* root):root(root){
+    }
+    CoNode* getRoot(){
+        return root;
+    }
+    void setRoot(CoNode* R){
+        root = R;
+    }
+
+};
+CoTree T;
+NetworKit::Graph G;
+int mark_count = 0;
+int mark_and_unmarked_count = 0;
+int mark_ever_count = 0;
+void unmark(queue<CoNode> &marked_with_d_equal_to_md){
```

Comment:
In general it's really misleading if you pass one parameter, but use 5 others because they are globals